### ISSUE TITLE
SDL2: direct3d batching causes strange issues on widescreen monitor

### ISSUE THREAD

**Issue by ivan-mogilko** [libsdl-org/SDL#7838]
According to the changelog, the "batching" mode was introduced in 2.0.10. In our project we were using a much newer version of SDL (started with 2.12 iirc, and 2.24 currently), and tbh I did never pay attention to these specifics. But recently I've been investigating a weird user's problem, and eventually found that it's related to this "batching" mode.

User has a following system (note the ultra-widescreen monitor):
- Windows 10 Pro
- Monitor: LG 34" WL500 (ultra-widescreen)
- NVIDIA GeForce RTX 3070 video card
- Desktop Resolution: 2560 x 1080 x 74 hertz

SDL_CreateRenderer seem to create a "direct3d" by default when we do not request particular one, and enables "batching" by default too. If we request "direct3d" explicitly, then the batching is off by default. (Btw, this discrepancy puzzled me a alot while investigating the problem, i even had to look into the library code to find out what's going on, as I was not aware of how this setting works.)

We tested setting SDL_HINT_RENDER_BATCHING to 1 and then tried other renderers (software, opengl), but only direct3d was causing the problem.

So the issue in question happens *only* with "direct3d" renderer and with batching on.

What happens is this: if the program starts in windowed mode, or fullscreen mode, both ways things look fine. If program is switched from fullscreen to windowed (or windowed -> fullscreen -> windowed), the window is never resized to a smaller size, instead it remains big covering almost whole desktop. However the game's image in the window look like it would be looking in windowed mode, aligned to a left side, or top-left corner.
Another thing that user noticed is that this "corrupt" window turns out to be "system topmost", other applications always stay below it.

Following are two videos recorded by user (he had to use a phone):
https://www.dropbox.com/s/aez61e8ezm1em4q/sdl-direct3dbatching.zip?dl=0
(archive is 18mb but i may attach to this ticket instead if you think that's better)

---

I'd like to find out if it's a problem in SDL2, or we are doing something obviously wrong in our program.

Without too much extra details, what we do there is:
```
SDL_SetWindowFullscreen(window, 0);
SDL_SetWindowResizable(window, SDL_TRUE);
```

followed by:

```
SDL_SetWindowFullscreen(window, 0);
SDL_SetWindowSize(window, width, height);
CenterWindow
SDL_SetWindowResizable(window, SDL_TRUE);
```
where "CenterWindow" is using code [posted in this comment](https://github.com/libsdl-org/SDL/issues/6875#issuecomment-1465347090), because SDL2 does not account for task bar when centering:
```
    SDL_Rect bounds;
    int w, h;
    int bx1, by1, bx2, by2;
    SDL_GetDisplayUsableBounds(SDL_GetWindowDisplayIndex(window), &bounds);
    SDL_GetWindowSize(window, &w, &h);
    SDL_GetWindowBordersSize(window, &by1, &bx1, &by2, &bx2);
    int x = bounds.x + bx1 + (bounds.w - (w + bx1 + bx2)) / 2;
    int y = bounds.y + by1 + (bounds.h - (h + by1 + by2)) / 2;
    SDL_SetWindowPosition(window, x, y);
```

As you may notice there's a certain repetition, where `SDL_SetWindowFullscreen(window, 0);` gets called twice eventually. This is an oversight in our code organization. But I cannot tell whether this may or not have anything to do with the issue.

**Comment by slouken** [comment]
We are scoping work for the SDL 3.2.0 release, so please let us know if this is a showstopper for you.

**Comment by ivan-mogilko** [comment]
> We are scoping work for the SDL 3.2.0 release, so please let us know if this is a showstopper for you.

No, it's not, this problem was reported by a single user so far, and it's possible to work around.

**Comment by ivan-mogilko** [comment]
An update for this, I am not sure whether I tested this inaccurately before, or this is because I have a different monitor now, but I have this issue on a pretty regular monitor now, and I have reports from another dev that he also experience same issue on his.

I may provide more information if needed.

Disabling batching prior to creating a renderer fixes the issue.
```SDL_SetHint(SDL_HINT_RENDER_BATCHING, "0");```

**Comment by slouken** [comment]
Are you seeing this with sdl2-compat and SDL3?

**Comment by ivan-mogilko** [comment]
I am using SDL 2.30.11 (from release-2.30.x branch). Tbh I don't know what "sdl2-compat" means. I haven't tried SDL3 yet.

**Comment by slouken** [comment]
Yes, it's an API compatibility layer using SDL3.

Go ahead and grab this and see if it works for you?
https://github.com/libsdl-org/sdl2-compat/releases/download/release-2.32.50/sdl2-compat-2.32.50-win32-x64.zip

**Comment by ivan-mogilko** [comment]
Window mode switching works well with SDL3 via SDL2 compat, with "direct3d" renderer (which was the offender here), but the problem is that apparently either the "batching" mode or setting is no longer a thing in SDL3: i cannot see the hint declared nor corresponding field in the internal SDL_Renderer struct.

Therefore, I am not sure if this issue is applicable to SDL3 at all.

**Comment by slouken** [comment]
Great, thanks for the confirmation. I would go with that, or one of your other workarounds for now.
